src/app/dim-ui/RichDestinyText.tsx

```diff
@@ -0,0 +1,109 @@
+  { icon: fusionRifle,     exampleObjectiveHash: 3296270292, unicode: '', substring: null },
+  { icon: smg,             exampleObjectiveHash: 2722409947, unicode: '', substring: null },
+  { icon: rocketLauncher,  exampleObjectiveHash: 2203404732, unicode: '', substring: null },
+  { icon: sidearm,         exampleObjectiveHash: 299893109,  unicode: '', substring: null },
+  { icon: melee,           exampleObjectiveHash: 3951261483, unicode: '', substring: null },
+  { icon: grenade,         exampleObjectiveHash: 3711356257, unicode: '', substring: null },
+  { icon: grenadeLauncher, exampleObjectiveHash: 2152699013, unicode: '', substring: null },
+  { icon: beamWeapon,      exampleObjectiveHash: 3080184954, unicode: '', substring: null },
+  { icon: damageSolar,     exampleObjectiveHash: 2994623161, unicode: '', substring: null },
+  { icon: damageKinetic,   exampleObjectiveHash: 3924246227, unicode: '', substring: null },
+  { icon: headshot,        exampleObjectiveHash: 3287913530, unicode: '', substring: null },
+  { icon: damageArc,       exampleObjectiveHash: 2178780271, unicode: '', substring: null },
+  { icon: damageVoid,      exampleObjectiveHash: 695106797,  unicode: '', substring: null },
+  { icon: wireRifle,       exampleObjectiveHash: 2923868331, unicode: '', substring: null },
+  { icon: swordHeavy,      exampleObjectiveHash: 989767424,  unicode: '', substring: null },
+  { icon: machinegun,      exampleObjectiveHash: 1788114534, unicode: '', substring: null }
+];
+
+const generateConversionTable = (defs) => {
+  // loop through conversionTable entries to update them with manifest string info
+  baseConversionTable.forEach((iconEntry) => {
+    const objectiveDef = defs.Objective.get(iconEntry.exampleObjectiveHash);
+    if (!objectiveDef) {
+      return;
+    }
+    // lookup this lang's string for the objective
+    const progressDescriptionMatch = objectiveDef.progressDescription.match(iconPlaceholder);
+    const iconString = progressDescriptionMatch && progressDescriptionMatch[0];
+    // this language's localized replacement, which we will detect and un-replace back into an icon
+    iconEntry.substring = iconString;
+  });
+  return baseConversionTable;
+};
+
+// returns the string-to-svg conversion table
+const generateConversionTableOnce = _.once(generateConversionTable);
+
+const replaceWithIcon = (
+  conversionRules: { icon: string; unicode: string; substring: string | null }[],
+  textSegment: string
+) => {
+  const replacement = conversionRules.find(
+    (r) => r.substring === textSegment || r.unicode === textSegment
+  );
+  return (
+    (replacement && <img src={replacement.icon} title={textSegment} key={textSegment} />) || (
+      <span key={textSegment}>{textSegment}</span>
+    )
+  );
+};
+
+export default function RichDestinyText({
+  text,
+  defs
+}: {
+  text: string;
+  defs?: D2ManifestDefinitions | D1ManifestDefinitions;
```

Comment:
D1 won't have any of these - they always used just text for all descriptions. We could add a `destinyVersion` property to the manifest definitions object that you could check if you wanted.